File: frontend/lib/marketplace.ts
export interface MarketplaceProductImage {
  id?: string | null;
  source_url: string;
  image_type: string;
  width?: number | null;
  height?: number | null;
  alt_text?: string | null;
  sort_order: number;
  is_primary: boolean;
}

export interface MarketplaceOffer {
  id?: string | null;
  provider_offer_id?: string | null;
  seller_name?: string | null;
  offer_title: string;
  offer_description?: string | null;
  price_minor?: number | null;
  price_inr?: number | null;
  currency: string;
  discount_minor?: number | null;
  discount_inr?: number | null;
  discount_percentage?: number | null;
  availability?: string | null;
  source_url?: string | null;
}

export interface MarketplaceFieldAvailability {
  has_original_description: boolean;
  has_images: boolean;
  has_offers: boolean;
  has_reviews: boolean;
  has_emi: boolean;
  has_seller_info: boolean;
  has_mrp: boolean;
  has_specifications: boolean;
}

export interface MarketplaceInternalMapping {
  mapping_status: "UNMAPPED" | "CANDIDATE" | "VERIFIED" | "DISABLED" | "PRICE_MISMATCH" | "UNAVAILABLE";
  mapping_confidence: number;
  internal_product_id?: string | null;
  internal_sku?: string | null;
  can_authoritative_checkout: boolean;
}

export interface MarketplaceProduct {
  id: string;
  provider: "amazon" | "flipkart" | "kharridlo_verified" | string;
  provider_product_id: string;
  canonical_url: string;
  title: string;
  brand: string;
  category: string;
  subcategory?: string | null;
  original_description?: string | null;
  normalized_description?: string | null;
  ai_summary?: string | null;
  specifications: Record<string, any>;
  source_currency: string;
  source_price_minor?: number | null;
  source_price_inr?: number | null;
  source_mrp_minor?: number | null;
  source_mrp_inr?: number | null;
  savings_inr?: number | null;
  availability_status: string;
  source_rating?: number | null;
  source_review_count?: number | null;
  seller_name?: string | null;
  images: MarketplaceProductImage[];
  primary_image_url?: string | null;
  offers: MarketplaceOffer[];
  review_summary?: any | null;
  finance_info?: any | null;
  field_availability: MarketplaceFieldAvailability;
  mapping?: MarketplaceInternalMapping | null;
  fetched_at: string;
}

export interface MarketplaceSearchResponse {
  items: MarketplaceProduct[];
  total: number;
  page: number;
  page_size: number;
  query?: string;
  providers_queried: string[];
  warnings: string[];
}

export interface MarketplaceProviderStatus {
  code: string;
  display_name: string;
  enabled: boolean;
  status: string;
  live_access_verified: boolean;
  notes?: string;
}

export function getProviderBadge(provider: string) {
  switch (provider) {
    case "amazon":
      return {
        label: "Amazon.in",
        provenance: "Provided by Amazon Creators API",
        badgeClass: "bg-amber-50 text-amber-900 border-amber-300 font-medium",
        dotClass: "bg-amber-500",
        isInternal: false,
      };
    case "flipkart":
      return {
        label: "Flipkart",
        provenance: "Provided by Flipkart Affiliate Feed",
        badgeClass: "bg-blue-50 text-blue-900 border-blue-300 font-medium",
        dotClass: "bg-blue-500",
        isInternal: false,
      };
    case "kharridlo_verified":
    default:
      return {
        label: "Kharridlo Verified",
        provenance: "Authoritative Kharridlo Escrow Catalog",
        badgeClass: "bg-emerald-50 text-emerald-900 border-emerald-300 font-semibold",
        dotClass: "bg-emerald-500",
        isInternal: true,
      };
  }
}

/**
 * Generates an authoritative, working external redirect link for Amazon & Flipkart.
 * Prevents broken 404 links by falling back to targeted keyword search if direct SKU/slug is unavailable.
 */
export function formatMarketplaceUrl(
  provider: string,
  productId?: string,
  title?: string,
  canonicalUrl?: string
): string {
  if (provider === "amazon") {
    const productAsin = productId?.match(/^[A-Z0-9]{10}$/i)?.[0];
    const canonicalAsin = canonicalUrl?.match(/\/dp\/([A-Z0-9]{10})(?:[/?]|$)/i)?.[1];
    const asin = productAsin || canonicalAsin;
    if (asin) return `https://www.amazon.in/dp/${asin.toUpperCase()}?tag=kharridlo-21`;
    const q = encodeURIComponent(title || productId || "student laptop");
    return `https://www.amazon.in/s?k=${q}&tag=kharridlo-21`;
  }

  if (provider === "flipkart") {
    if (canonicalUrl) {
      try {
        const url = new URL(canonicalUrl);
        const isFlipkart = url.hostname === "flipkart.com" || url.hostname.endsWith(".flipkart.com");
        const isProductPage = /\/p\/itm[a-z0-9]{8,}/i.test(url.pathname);
        if (isFlipkart && isProductPage) {
          url.searchParams.set("affid", "kharridlo");
          return url.toString();
        }
      } catch {
        // An invalid or incomplete canonical URL is intentionally replaced by a safe search URL.
      }
    }
    const q = encodeURIComponent(title || productId || "electronics");
    return `https://www.flipkart.com/search?q=${q}&affid=kharridlo`;
  }

  return canonicalUrl || `/product/${productId || ""}`;
}

// Backwards-compatible name used by existing product pages.
export const getMarketplaceRedirectUrl = formatMarketplaceUrl;

File: backend/app/schemas/marketplace.py
from typing import Optional, List, Dict, Any
from datetime import datetime, timezone
from pydantic import BaseModel, ConfigDict, Field, computed_field


class MarketplaceProductImageResponse(BaseModel):
    id: Optional[str] = None
    source_url: str
    image_type: str = "gallery"  # 'primary', 'thumbnail', 'gallery'
    width: Optional[int] = None
    height: Optional[int] = None
    alt_text: Optional[str] = None
    sort_order: int = 0
    is_primary: bool = False

    model_config = ConfigDict(from_attributes=True)


class MarketplaceOfferResponse(BaseModel):
    id: Optional[str] = None
    provider_offer_id: Optional[str] = None
    seller_name: Optional[str] = None
    offer_title: str
    offer_description: Optional[str] = None
    price_minor: Optional[int] = None
    currency: str = "INR"
    discount_minor: Optional[int] = None
    discount_percentage: Optional[float] = None
    availability: Optional[str] = None
    source_url: Optional[str] = None

    @computed_field
    @property
    def price_inr(self) -> Optional[float]:
        return round(self.price_minor / 100.0, 2) if self.price_minor is not None else None

    @computed_field
    @property
    def discount_inr(self) -> Optional[float]:
        return round(self.discount_minor / 100.0, 2) if self.discount_minor is not None else None

    model_config = ConfigDict(from_attributes=True)


class MarketplaceReviewSummaryResponse(BaseModel):
    rating: float
    review_count: int
    rating_distribution: Dict[str, Any] = Field(default_factory=dict)
    source: str
    fetched_at: datetime

    model_config = ConfigDict(from_attributes=True)


class MarketplaceFinanceResponse(BaseModel):
    emi_available: bool = False
    emi_provider: Optional[str] = None
    emi_summary: Optional[str] = None
    no_cost_emi_available: bool = False
    source_text: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class MarketplaceFieldAvailability(BaseModel):
    """Honest tracking of whether fields were authentic vs unavailable from provider."""
    has_original_description: bool = False
    has_images: bool = False
    has_offers: bool = False
    has_reviews: bool = False
    has_emi: bool = False
    has_seller_info: bool = False
    has_mrp: bool = False
    has_specifications: bool = False


class MarketplaceInternalMappingResponse(BaseModel):
    mapping_status: str = "UNMAPPED"  # 'UNMAPPED', 'CANDIDATE', 'VERIFIED', 'DISABLED', 'PRICE_MISMATCH', 'UNAVAILABLE'
    mapping_confidence: float = 0.0
    internal_product_id: Optional[str] = None
    internal_sku: Optional[str] = None
    can_authoritative_checkout: bool = False

    model_config = ConfigDict(from_attributes=True)


class MarketplaceProductResponse(BaseModel):
    id: str
    provider: str  # 'amazon', 'flipkart', 'kharridlo_verified'
    provider_product_id: str
    canonical_url: str
    title: str
    brand: str
    category: str
    subcategory: Optional[str] = None

    # Provenance separation
    original_description: Optional[str] = None
    normalized_description: Optional[str] = None
    ai_summary: Optional[str] = None
    specifications: Dict[str, Any] = Field(default_factory=dict)

    # Pricing
    source_currency: str = "INR"
    source_price_minor: Optional[int] = None
    source_mrp_minor: Optional[int] = None
    availability_status: str = "in_stock"

    # Ratings & Seller
    source_rating: Optional[float] = None
    source_review_count: Optional[int] = None
    seller_name: Optional[str] = None

    # Rich relations
    images: List[MarketplaceProductImageResponse] = Field(default_factory=list)
    primary_image_url: Optional[str] = None
    offers: List[MarketplaceOfferResponse] = Field(default_factory=list)
    review_summary: Optional[MarketplaceReviewSummaryResponse] = None
    finance_info: Optional[MarketplaceFinanceResponse] = None

    # Gating & Provenance metadata
    fetched_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    field_availability: MarketplaceFieldAvailability = Field(default_factory=MarketplaceFieldAvailability)
    mapping: Optional[MarketplaceInternalMappingResponse] = None

    @computed_field
    @property
    def source_price_inr(self) -> Optional[float]:
        return round(self.source_price_minor / 100.0, 2) if self.source_price_minor is not None else None

    @computed_field
    @property
    def source_mrp_inr(self) -> Optional[float]:
        return round(self.source_mrp_minor / 100.0, 2) if self.source_mrp_minor is not None else None

    @computed_field
    @property
    def savings_inr(self) -> Optional[float]:
        if self.source_mrp_minor is not None and self.source_price_minor is not None:
            diff = self.source_mrp_minor - self.source_price_minor
            return round(diff / 100.0, 2) if diff > 0 else 0.0
        return None

    model_config = ConfigDict(from_attributes=True)


class MarketplaceSearchResponse(BaseModel):
    items: List[MarketplaceProductResponse] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 20
    query: Optional[str] = None
    providers_queried: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)


class MarketplaceProviderStatusResponse(BaseModel):
    code: str
    display_name: str
    enabled: bool
    status: str  # 'active', 'degraded', 'unavailable', 'rate_limited'
    live_access_verified: bool = False
    last_success_at: Optional[datetime] = None
    last_error_at: Optional[datetime] = None
    last_error_message: Optional[str] = None
    notes: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)
